Bottom: 79b36b9ed22f77cfe81e888155e383bcf09e5e9f
Top:    9023ede0dfcb282ab90f7726e5c637c9ff842534
Author: Sean Hefty <sean.hefty@intel.com>
Date:   2012-06-05 16:58:52 -0700

Refresh of rs-spin

---

diff --git a/src/cma.c b/src/cma.c
index 140e22c..f688c4c 100755
--- a/src/cma.c
+++ b/src/cma.c
@@ -276,7 +276,6 @@ int ucma_init(void)
 
 	cma_dev_cnt = dev_cnt;
 	ucma_set_af_ib_support();
-	rs_configure();
 	pthread_mutex_unlock(&mut);
 	ibv_free_device_list(dev_list);
 	return 0;
diff --git a/src/cma.h b/src/cma.h
index 828c6aa..cedc0c3 100644
--- a/src/cma.h
+++ b/src/cma.h
@@ -114,7 +114,6 @@ static inline int ERR(int err)
 }
 
 int ucma_init();
-void rs_configure();
 extern int af_ib_support;
 
 #define RAI_ROUTEONLY		0x01000000
diff --git a/src/rsocket.c b/src/rsocket.c
index bb66503..90f9a10 100644
--- a/src/rsocket.c
+++ b/src/rsocket.c
@@ -204,11 +204,22 @@ struct rsocket {
 void rs_configure(void)
 {
 	FILE *f;
+	static int init;
+
+	if (init)
+		return;
+
+	pthread_mutex_lock(&mut);
+	if (init)
+		goto out;
 
 	if ((f = fopen(RS_CONF_DIR "/polling_time", "r"))) {
 		fscanf(f, "%u", &polling_time);
 		fclose(f);
 	}
+	init = 1;
+out:
+	pthread_mutex_unlock(&mut);
 }
 
 /*
@@ -485,6 +496,7 @@ int rsocket(int domain, int type, int protocol)
 	    (type != SOCK_STREAM) || (protocol && protocol != IPPROTO_TCP))
 		return ERR(ENOTSUP);
 
+	rs_configure();
 	rs = rs_alloc(NULL);
 	if (!rs)
 		return ERR(ENOMEM);
